compiler/executable/match_/planner/mod.rs

```diff
@@ -307,16 +307,23 @@ impl MatchExecutableBuilder {
     }
 
     fn push_check(&mut self, variables: &[Variable], check: CheckInstruction<ExecutorVariable>) {
+        // if it is a comparison or IID (TODO) we can inline the check into previous instructions
+        self.inline_as_optimisation(variables, &check);
+
+        // try to inline the check into the currently-built intersection, instead of creating a separate step
         if let Some(intersection) = self.current.as_mut().and_then(|b| b.builder.as_intersection_mut()) {
+            let mut is_added = false;
             for instruction in intersection.instructions.iter_mut().rev() {
-                let mut is_producer = false;
-                instruction.new_variables_foreach(|var| is_producer |= variables.contains(&self.reverse_index[&var]));
+                let is_producer = variables.iter().any(|var| instruction.is_new_variable(self.index[var]));
                 if is_producer {
-                    instruction.add_check(check);
-                    self.current.as_mut().unwrap().selected_variables = self.current_outputs.clone();
-                    return;
+                    instruction.add_check(check.clone());
+                    is_added = true;
                 }
             }
+            if is_added {
+                self.current.as_mut().unwrap().selected_variables = self.current_outputs.clone();
+                return;
+            }
         }
```

Comment:
Can this be unified with the body of `inline_as_optimisation()`? The code isn't quite the same, but it presumably does roughly the same thing?